### ISSUE TITLE
App crashes every ~10 votes 

### ISSUE THREAD

**Issue by lappalj4** [anishathalye/gavel#35]
I deployed Gavel to Heroku on a hobby dyno, and I'm seeing a ton of crashes when voting on items. We had around 170 items and 10 annotators spamming the app simultaneously, and the app would show an Internal Server Error quite regularly after voting.

I checked the Heroku logs and this is the error that keeps repeating: 

> 2016-11-07T17:45:24.292366+00:00 app[web.1]: [2016-11-07 17:45:24,289] ERROR in app: Exception on /vote [POST]
2016-11-07T17:45:24.292378+00:00 app[web.1]: Traceback (most recent call last):
2016-11-07T17:45:24.292380+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
2016-11-07T17:45:24.292382+00:00 app[web.1]:     context)
2016-11-07T17:45:24.292383+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
2016-11-07T17:45:24.292384+00:00 app[web.1]:     cursor.execute(statement, parameters)
2016-11-07T17:45:24.292385+00:00 app[web.1]: psycopg2.extensions.TransactionRollbackError: could not serialize access due to concurrent update
2016-11-07T17:45:24.292386+00:00 app[web.1]:
2016-11-07T17:45:24.292387+00:00 app[web.1]:
2016-11-07T17:45:24.292387+00:00 app[web.1]: The above exception was the direct cause of the following exception:
2016-11-07T17:45:24.292388+00:00 app[web.1]:
2016-11-07T17:45:24.292389+00:00 app[web.1]: Traceback (most recent call last):
2016-11-07T17:45:24.292390+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/app.py", line 1988, in wsgi_app
2016-11-07T17:45:24.292391+00:00 app[web.1]:     response = self.full_dispatch_request()
2016-11-07T17:45:24.292392+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/app.py", line 1641, in full_dispatch_request
2016-11-07T17:45:24.292392+00:00 app[web.1]:     rv = self.handle_user_exception(e)
2016-11-07T17:45:24.292393+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/app.py", line 1544, in handle_user_exception
2016-11-07T17:45:24.292394+00:00 app[web.1]:     reraise(exc_type, exc_value, tb)
2016-11-07T17:45:24.292420+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
2016-11-07T17:45:24.292422+00:00 app[web.1]:     raise value
2016-11-07T17:45:24.292422+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/app.py", line 1639, in full_dispatch_request
2016-11-07T17:45:24.292423+00:00 app[web.1]:     rv = self.dispatch_request()
2016-11-07T17:45:24.292424+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/flask/app.py", line 1625, in dispatch_request
2016-11-07T17:45:24.292425+00:00 app[web.1]:     return self.view_functions[rule.endpoint](**req.view_args)
2016-11-07T17:45:24.292426+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 25, in decorated
2016-11-07T17:45:24.292427+00:00 app[web.1]:     return f(*args, **kwargs)
2016-11-07T17:45:24.292428+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 33, in decorated
2016-11-07T17:45:24.292429+00:00 app[web.1]:     annotator = get_current_annotator()
2016-11-07T17:45:24.292429+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 150, in get_current_annotator
2016-11-07T17:45:24.292430+00:00 app[web.1]:     return Annotator.by_id(session.get(ANNOTATOR_ID, None))
2016-11-07T17:45:24.292430+00:00 app[web.1]:   File "/app/gavel/models/annotator.py", line 59, in by_id
2016-11-07T17:45:24.292431+00:00 app[web.1]:     annotator = cls.query.with_for_update().get(uid)
2016-11-07T17:45:24.292432+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 831, in get
2016-11-07T17:45:24.292432+00:00 app[web.1]:     return self._get_impl(ident, loading.load_on_ident)
2016-11-07T17:45:24.292433+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 864, in _get_impl
2016-11-07T17:45:24.292433+00:00 app[web.1]:     return fallback_fn(self, key)
2016-11-07T17:45:24.292434+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 219, in load_on_ident
2016-11-07T17:45:24.292434+00:00 app[web.1]:     return q.one()
2016-11-07T17:45:24.292435+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2718, in one
2016-11-07T17:45:24.292436+00:00 app[web.1]:     ret = list(self)
2016-11-07T17:45:24.292436+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2761, in __iter__
2016-11-07T17:45:24.292437+00:00 app[web.1]:     return self._execute_and_instances(context)
2016-11-07T17:45:24.292438+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2776, in _execute_and_instances
2016-11-07T17:45:24.292439+00:00 app[web.1]:     result = conn.execute(querycontext.statement, self._params)
2016-11-07T17:45:24.292439+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 914, in execute
2016-11-07T17:45:24.292440+00:00 app[web.1]:     return meth(self, multiparams, params)
2016-11-07T17:45:24.292441+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
2016-11-07T17:45:24.292441+00:00 app[web.1]:     return connection._execute_clauseelement(self, multiparams, params)
2016-11-07T17:45:24.292442+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
2016-11-07T17:45:24.292443+00:00 app[web.1]:     compiled_sql, distilled_params
2016-11-07T17:45:24.292443+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
2016-11-07T17:45:24.292444+00:00 app[web.1]:     context)
2016-11-07T17:45:24.292445+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
2016-11-07T17:45:24.292445+00:00 app[web.1]:     exc_info
2016-11-07T17:45:24.292446+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 202, in raise_from_cause
2016-11-07T17:45:24.292447+00:00 app[web.1]:     reraise(type(exception), exception, tb=exc_tb, cause=cause)
2016-11-07T17:45:24.292448+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 185, in reraise
2016-11-07T17:45:24.292448+00:00 app[web.1]:     raise value.with_traceback(tb)
2016-11-07T17:45:24.292449+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
2016-11-07T17:45:24.292450+00:00 app[web.1]:     context)
2016-11-07T17:45:24.292450+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
2016-11-07T17:45:24.292457+00:00 app[web.1]:     cursor.execute(statement, parameters)
2016-11-07T17:45:24.292459+00:00 app[web.1]: sqlalchemy.exc.OperationalError: (psycopg2.extensions.TransactionRollbackError) could not serialize access due to concurrent update
2016-11-07T17:45:24.292548+00:00 app[web.1]:  [SQL: 'SELECT annotator.id AS annotator_id, annotator.name AS annotator_name, annotator.email AS annotator_email, annotator.active AS annotator_active, annotator.read_welcome AS annotator_read_welcome, annotator.description AS annotator_description, annotator.secret AS annotator_secret, annotator.next_id AS annotator_next_id, annotator.updated AS annotator_updated, annotator.prev_id AS annotator_prev_id, annotator.alpha AS annotator_alpha, annotator.beta AS annotator_beta \nFROM annotator \nWHERE annotator.id = %(param_1)s FOR UPDATE'] [parameters: {'param_1': 6}]

Could this be an issue with limits in the hobby dyno or something else? Would be nice to know that this won't be an issue on game day in three weeks :D

**Comment by anishathalye** [comment]
It's not an issue with the hobby dyno.

Gavel sets the database isolation level to serialized (see: https://github.com/anishathalye/gavel/blob/master/gavel/models/__init__.py#L8). We choose this consistency model (see here for some background: https://aphyr.com/posts/313-strong-consistency-models) because we don't want weird situations to arise. For example, we don't want to get a bunch of values from the database, then do some computation, and then write those values back while they might have been concurrently modified by another process.

You see that Internal Server Error whenever operations interleave in a "bad" way. The way the driver handles that is that the interleaved operations just don't commit (to preserve database integrity).

This should be handled better than it is currently.

One simple way to handle this is to add some code around the `db.session.commit()` code, and when a TransactionRollbackError occurs, just don't do anything and reload `/`. The judges will see the same content again, and they should just click on the same button again.

A better way to handle this could be to retry the entire operation for a fixed number of times (until it succeeds), otherwise display some error (or follow the strategy mentioned above).

In any case, I don't think this comes up that often in practice. At least at HackMIT 2016, no judge mentioned this problem to us. And if it does occur, judges will probably just click the "back" button in their browser and retry. So it should definitely be fixed at some point, but it shouldn't be a huge issue if it's not fixed before your event.

**Comment by lappalj4** [comment]
Alright, thanks for the exhaustive reply :)

**Comment by lappalj4** [comment]
Whoops, maybe it's best that this stays open since it could as well be improved

**Comment by anishathalye** [comment]
Yeah, definitely. If you're interested in contributing a fix, that would be awesome. Otherwise, someone on the HackMIT team should get to it eventually.

**Comment by theikkila** [comment]
The consistency model is wrong now, even though the current implementation of saving the scores of judges is little awkward by mutating and not ie. adding just the rating into database as immutable scores.

Serializable transaction isolation should be used if there is concurrent transactions in same table that must be reliably resolved.

Lets take an concrete example:

Here is table `results`:
![image](https://cloud.githubusercontent.com/assets/608642/20647271/b87d8f50-b498-11e6-9054-a7f18d62201e.png)


![image](https://cloud.githubusercontent.com/assets/608642/20647235/971fc9dc-b497-11e6-9511-6be332e04177.png)
Virtually both transactions are happening concurrently and so with serializable-level isolation the consistency cannot be achieved and the other transaction will fail.

With read commited isolation this both transactions will get the state of the database like it was in same instant than when the transaction started:
![image](https://cloud.githubusercontent.com/assets/608642/20647299/8fad9920-b499-11e6-8d47-5203285558e9.png)
Now the data may be corrupted because even the first transaction is commited before the second its not returning transaction error but valid state.

I cannot be 100% sure that this more lightweight isolation would not cause any problems but note that the usecase of this application shouldn't sacrifice availability for absolute consistency.

**Comment by anishathalye** [comment]
I think the database isolation levels provide guarantees at the database level, not just at the table level.

Yeah, the current design of mutable data is not great, but given that design, we need strong isolation guarantees. A vote from a judge reads from the judge's row and the rows for both items, and it updates all three of those rows it read from (actually, it does more than that, but we can ignore the other stuff for now).

Looking at the [transaction isolation guarantees](https://www.postgresql.org/docs/9.1/static/transaction-iso.html), "any concurrent execution of a set of Serializable transactions is guaranteed to produce the same effect as running them one at a time in some order" (and some transactions may fail to commit). At the very least, this should produce correct results for our use case.

The worst thing that happens when there are incompatible concurrent updates is that the judges have to refresh the page to try again (and this happens very rarely in real-world use, in my experience). For a judging system, we need absolute correctness/consistency, even if that means some inconvenience for judges once in a while.

Anyways, ideally, we change the thing so that votes are append-only immutable values and rankings are computed on the fly. Also, this would make #22 a whole lot easier.

**Comment by anishathalye** [comment]
Turns out that there are some events that are large enough / have enough concurrent judge input that this is an issue (e.g. Junction, which had > 300 projects). This should be fixed.

**Comment by charliealbright** [comment]
@anishathalye We are using Gavel for a hackathon with ~100 teams and ~80 judges - during preliminary testing with about 20 sample teams and 20 sample judges, we have been seeing this error a lot. I'm trying to work on a quick fix but not sure if I'll have it done by the time of our event.

Here are some more Heroku logs you can look at if it helps:

```
2018-04-19T03:00:00.522334+00:00 app[web.1]: [2018-04-19 03:00:00,520] ERROR in app: Exception on /vote [POST]
2018-04-19T03:00:00.522338+00:00 app[web.1]: Traceback (most recent call last):
2018-04-19T03:00:00.522340+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
2018-04-19T03:00:00.522343+00:00 app[web.1]:     context)
2018-04-19T03:00:00.522345+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
2018-04-19T03:00:00.522347+00:00 app[web.1]:     cursor.execute(statement, parameters)
2018-04-19T03:00:00.522349+00:00 app[web.1]: psycopg2.extensions.TransactionRollbackError: could not serialize access due to read/write dependencies among transactions
2018-04-19T03:00:00.522351+00:00 app[web.1]: DETAIL:  Reason code: Canceled on identification as a pivot, during conflict out checking.
2018-04-19T03:00:00.522353+00:00 app[web.1]: HINT:  The transaction might succeed if retried.
2018-04-19T03:00:00.522355+00:00 app[web.1]: 
2018-04-19T03:00:00.522356+00:00 app[web.1]: 
2018-04-19T03:00:00.522358+00:00 app[web.1]: The above exception was the direct cause of the following exception:
2018-04-19T03:00:00.522360+00:00 app[web.1]: 
2018-04-19T03:00:00.522362+00:00 app[web.1]: Traceback (most recent call last):
2018-04-19T03:00:00.522364+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
2018-04-19T03:00:00.522365+00:00 app[web.1]:     response = self.full_dispatch_request()
2018-04-19T03:00:00.522367+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
2018-04-19T03:00:00.522369+00:00 app[web.1]:     rv = self.handle_user_exception(e)
2018-04-19T03:00:00.522371+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
2018-04-19T03:00:00.522372+00:00 app[web.1]:     reraise(exc_type, exc_value, tb)
2018-04-19T03:00:00.522374+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
2018-04-19T03:00:00.522376+00:00 app[web.1]:     raise value
2018-04-19T03:00:00.522377+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
2018-04-19T03:00:00.522379+00:00 app[web.1]:     rv = self.dispatch_request()
2018-04-19T03:00:00.522381+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
2018-04-19T03:00:00.522383+00:00 app[web.1]:     return self.view_functions[rule.endpoint](**req.view_args)
2018-04-19T03:00:00.522384+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 25, in decorated
2018-04-19T03:00:00.522386+00:00 app[web.1]:     return f(*args, **kwargs)
2018-04-19T03:00:00.522388+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 37, in decorated
2018-04-19T03:00:00.522390+00:00 app[web.1]:     return f(*args, **kwargs)
2018-04-19T03:00:00.522391+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 95, in vote
2018-04-19T03:00:00.522393+00:00 app[web.1]:     annotator.update_next(choose_next(annotator))
2018-04-19T03:00:00.522395+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 193, in choose_next
2018-04-19T03:00:00.522397+00:00 app[web.1]:     items = preferred_items(annotator)
2018-04-19T03:00:00.522398+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 181, in preferred_items
2018-04-19T03:00:00.522400+00:00 app[web.1]:     less_seen = [i for i in preferred if len(i.viewed) < settings.MIN_VIEWS]
2018-04-19T03:00:00.522402+00:00 app[web.1]:   File "/app/gavel/controllers/judge.py", line 181, in <listcomp>
2018-04-19T03:00:00.522403+00:00 app[web.1]:     less_seen = [i for i in preferred if len(i.viewed) < settings.MIN_VIEWS]
2018-04-19T03:00:00.522405+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/attributes.py", line 237, in __get__
2018-04-19T03:00:00.522407+00:00 app[web.1]:     return self.impl.get(instance_state(instance), dict_)
2018-04-19T03:00:00.522408+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/attributes.py", line 584, in get
2018-04-19T03:00:00.522410+00:00 app[web.1]:     value = self.callable_(state, passive)
2018-04-19T03:00:00.522412+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/strategies.py", line 557, in _load_for_state
2018-04-19T03:00:00.522414+00:00 app[web.1]:     return self._emit_lazyload(session, state, ident_key, passive)
2018-04-19T03:00:00.522415+00:00 app[web.1]:   File "<string>", line 1, in <lambda>
2018-04-19T03:00:00.522417+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/strategies.py", line 635, in _emit_lazyload
2018-04-19T03:00:00.522419+00:00 app[web.1]:     result = q.all()
2018-04-19T03:00:00.522421+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2679, in all
2018-04-19T03:00:00.522422+00:00 app[web.1]:     return list(self)
2018-04-19T03:00:00.522424+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2831, in __iter__
2018-04-19T03:00:00.522425+00:00 app[web.1]:     return self._execute_and_instances(context)
2018-04-19T03:00:00.522427+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2854, in _execute_and_instances
2018-04-19T03:00:00.522429+00:00 app[web.1]:     result = conn.execute(querycontext.statement, self._params)
2018-04-19T03:00:00.522431+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 945, in execute
2018-04-19T03:00:00.522432+00:00 app[web.1]:     return meth(self, multiparams, params)
2018-04-19T03:00:00.522434+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
2018-04-19T03:00:00.522436+00:00 app[web.1]:     return connection._execute_clauseelement(self, multiparams, params)
2018-04-19T03:00:00.522437+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
2018-04-19T03:00:00.522439+00:00 app[web.1]:     compiled_sql, distilled_params
2018-04-19T03:00:00.522441+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
2018-04-19T03:00:00.522442+00:00 app[web.1]:     context)
2018-04-19T03:00:00.522449+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1393, in _handle_dbapi_exception
2018-04-19T03:00:00.522451+00:00 app[web.1]:     exc_info
2018-04-19T03:00:00.522453+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
2018-04-19T03:00:00.522454+00:00 app[web.1]:     reraise(type(exception), exception, tb=exc_tb, cause=cause)
2018-04-19T03:00:00.522456+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
2018-04-19T03:00:00.522458+00:00 app[web.1]:     raise value.with_traceback(tb)
2018-04-19T03:00:00.522459+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
2018-04-19T03:00:00.522461+00:00 app[web.1]:     context)
2018-04-19T03:00:00.522462+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
2018-04-19T03:00:00.522464+00:00 app[web.1]:     cursor.execute(statement, parameters)
2018-04-19T03:00:00.522466+00:00 app[web.1]: sqlalchemy.exc.OperationalError: (psycopg2.extensions.TransactionRollbackError) could not serialize access due to read/write dependencies among transactions
2018-04-19T03:00:00.522468+00:00 app[web.1]: DETAIL:  Reason code: Canceled on identification as a pivot, during conflict out checking.
2018-04-19T03:00:00.522470+00:00 app[web.1]: HINT:  The transaction might succeed if retried.
```

It almost seems like everything in the `vote():` function in `judge.py` needs to be able to catch a db exception, then retry. Something like:

```
try:
    # do stuff
    db.session.commit()
except:
    db.session.rollback()
    # maybe wait here?
    vote()
```

**Comment by anishathalye** [comment]
@charliealbright this is a known behavior (of the [serializable isolation](https://www.postgresql.org/docs/current/static/transaction-iso.html#XACT-SERIALIZABLE) level, which is necessary for correctness), and a retry loop would be a good thing to have here.

In practice, though, this hasn't really been much of a problem (lots of events have used this without issue). Concurrent transactions don't really happen in practice all that often (a transaction may take a second to run, which is fast compared to the rate at which judges vote on projects).

Is your preliminary testing based on simulating a bunch of judges in software (or something similar, which will have people voting much faster than they would in practice)? Also, when is your event?

**Comment by charliealbright** [comment]
@anishathalye Our event is today - I ended up writing in some retry logic and the error rate dropped significantly, but is not completely eliminated. The fix basically retries DB transactions until they succeed, so it's not a long term solution, but it'll work for this event. 

There are ~100 teams and 65 verified judges, and we are running it over the course of 2 hours. I'll post an update sometime later after I have time to analyze the data.

During testing, we were voting pretty frequently, so I'm hoping there are virtually no errors during the real world conditions of voting in the venue.

**Comment by anishathalye** [comment]
It's actually a pretty okay long-term solution too 😛 

Feel free to send a pull request after your event when you have some free time.

And hope everything goes well with the event!

**Comment by charliealbright** [comment]
Thanks! It's going very well so far, basically no errors. However, I think there may be one small issue with my fix - 3 or 4 judges have experienced an error, and when they navigated back and refreshed, the current/previous teams were 2 they haven't seen before. So maybe I can dig into that or put up a PR and get more eyes on it.

**Comment by anishathalye** [comment]
Retry logic implemented in e3c2fb0f2a80f9544ded49b3c7fe0f39ab128dc7. Users shouldn't see this error anymore.
